src/ILCompiler.Compiler/src/CppCodeGen/CppWriter.cs

```diff
@@ -770,7 +770,8 @@ private String GetCodeForObjectNode(ObjectNode node, NodeFactory factory)
             }
             else
             {
-                nodeCode.Append(node.GetName());
+                // Rename generic composition nodes to avoid name clash with types
+                nodeCode.Append(node is GenericCompositionNode ? node.GetName().Replace("::", "_") : node.GetName());
```

Comment:
`GetName()` is just a `ToString()` (we should probably rename it as such). There shouldn't be any code based of off this since it has no contractual requirements. We mostly let it fallback to whatever `MangledName` would return, but that's just our laziness.

Can you use `ISymbol.MangledName`?